typescript/sdk/src/ism/HyperlaneIsmFactory.ts

```diff
@@ -0,0 +1,250 @@
+        config.threshold,
+      );
+      matches = eqAddress(expectedAdddress, module.address);
+      break;
+    }
+    case ModuleType.ROUTING: {
+      // A RoutingIsm matches if:
+      //   1. The set of domains in the config equals those on-chain
+      //   2. The modules for each domain match the config
+      // TODO: Check (1)
+      const routingIsm = DomainRoutingIsm__factory.connect(
+        moduleAddress,
+        provider,
+      );
+      const owner = await routingIsm.owner();
+      matches = matches && eqAddress(owner, config.owner);
+      for (const chain of Object.keys(config.domains)) {
+        const subModule = await routingIsm.modules(
+          multiProvider.getDomainId(chain),
+        );
+        const subModuleMatches = await moduleMatches(
+          chain,
+          subModule,
+          config.domains[chain],
+          multiProvider,
+          contracts,
+        );
+        matches = matches && subModuleMatches;
+      }
+      break;
+    }
+    case ModuleType.AGGREGATION: {
+      // An AggregationIsm matches if:
+      //   1. The threshold matches the config
+      //   2. There is a bijection between on and off-chain configured modules
+      const aggregationIsm = StaticAggregationIsm__factory.connect(
+        moduleAddress,
+        provider,
+      );
+      const [modules, threshold] = await aggregationIsm.modulesAndThreshold(
+        '0x',
+      );
+      matches = matches && threshold === config.threshold;
+      matches = matches && modules.length === config.modules.length;
+      const matched = config.modules.map(() => false);
+      for (const subModule of modules) {
+        const matching = await Promise.all(
+          config.modules.map((c) =>
+            moduleMatches(chain, subModule, c, multiProvider, contracts),
+          ),
+        );
+        const count = matching.filter((x) => x).length;
+        matches = matches && count === 1;
+        matched[matching.indexOf(true)] = true;
+      }
+      matches = matches && matched.every(Boolean);
+      break;
+    }
+    default: {
+      throw new Error('Unsupported ModuleType');
+    }
+  }
+
+  return matches;
+}
+
+export function collectValidators(
+  origin: ChainName,
+  config: IsmConfig,
+): Set<string> {
+  const validators = new Set<string>();
+
+  switch (config.type) {
+    case ModuleType.MULTISIG: {
+      config.validators.map((v) => validators.add(v));
+      break;
+    }
+    case ModuleType.ROUTING: {
+      if (Object.keys(config.domains).includes(origin)) {
+        const domainValidators = collectValidators(
+          origin,
+          config.domains[origin],
+        );
+        [...domainValidators].map((v) => validators.add(v));
+      }
+      break;
+    }
+    case ModuleType.AGGREGATION: {
+      const aggregatedValidators = config.modules.map((c) =>
+        collectValidators(origin, c),
+      );
+      aggregatedValidators.map((set) => {
+        [...set].map((v) => validators.add(v));
```

Comment:
alternatively instead of building the set iteratively, get the list and then return new Set(list)

Reply:
esp with the switch case the fallthrough is a little scary